test/emulator_backend.go

```diff
@@ -183,12 +183,7 @@ func (e *EmulatorBackend) RunScript(
 	}
 
 	staticType := runtime.ImportType(inter, result.Value.Type())
-	expectedType, err := inter.ConvertStaticToSemaType(staticType)
-	if err != nil {
-		return &stdlib.ScriptResult{
-			Error: err,
-		}
-	}
```

Comment:
Was this change intentional? The returned error would still need to be handled

Reply:
Yes, this change is indeed intentional. Let me explain why.

Suppose that a script has the return value: `[A.f8d6e0586b0a20c7.FooContract.SpecialNumber]`

The `staticType` variable from calling `runtime.ImportType(inter, result.Value.Type())`,
would be: `*cadence.VariableSizedArrayType`

Converting that `staticType` to the `expectedType`, with: `inter.ConvertStaticToSemaType(staticType)`,
would return the following error: `message: "failed to load type: "` and the `expectedType` would be `nil`.
I am guessing because this is actually a run-time type, and not a static type?

However, if we ignore the resulting error, calling `runtime.ImportValue(...)`, with a `nil` `expectedType`, works just fine.

I have to be honest, I do not fully understand why :sweat_smile: But it works :man_shrugging: 

Do you have any idea why would that be?

Reply:
> ... calling runtime.ImportValue(...), with a nil expectedType, works just fine.

This behavior is correct/expected. The reason is, `ImportValue` function accepts the `expectedType`  to only infer the type where necessary. For e.g. consider a case where an array of values `[2, 4, 6]` is passed. If you don't pass any expected type (or `nil`), the type for the resulting array would be inferred as `[Int]`. But if you pass `[UInt8]` as the `expectedType`, then the type for the decoded array value will also be `[UInt8]`.

Reply:
So the danger of ignoring the error is also the same as above. Imagine the returned array from the script is of type `[SomeStructInterface]`. Imagine there was an error loading the `SomeStructInterface` type, but we continue to import the value. 

By any chance, if the actual values in the array are of different types `[Foo(), Bar()]` where `Foo` and `Bar` implement `SomeStructInterface`, and also other common interfaces, then the inferred type is not guaranteed to be the same as `[SomeStructInterface]`. It would be `[T]` where `T` could be any of `AnyStruct`, `SomeStructInterface`, or even some other interface that is common to both `Foo` and `Bar`.

Reply:
> Converting that staticType to the expectedType, with: inter.ConvertStaticToSemaType(staticType),
would return the following error: message: "failed to load type: " and the expectedType would be nil.
I am guessing because this is actually a run-time type, and not a static type?

That's a bit strange. In both `inter.ConvertStaticToSemaType(staticType)` and in `ImportValue`, they use the same `inter.GetCompositeType(location, qualifiedIdentifier, typeID)` function to import/convert the type. Can you maybe check why it works in one but fails in the other?

Reply:
Yes, I actually tracked down why & where that happens. It is this method: https://github.com/onflow/cadence/blob/v0.39.14/runtime/convertTypes.go#L528-L535.

However, this seems to have been updated in master: https://github.com/onflow/cadence/blob/52549529e96b6094bb726e5e50d63f34e2eb292b/runtime/convertTypes.go#L528-L534

I am guessing that must have been fixed upstream, but yet not included in the Cadence version that the testing framework uses.

Reply:
@SupunS I checked this with the latest master branch of Cadence, and this is indeed fixed by a recent commit from @turbolent (see here: https://github.com/onflow/cadence/pull/2751/files#diff-b8dbb829a0ee92b4059eee848ee82a187d37f4115232586ed2b00ab4fe3eaf2fR529).

I will update the code to handle the returned error from `inter.ConvertStaticToSemaType`.

Reply:
Nice!